auth/scoperesolver.js

```diff
@@ -0,0 +1,338 @@
+          throw new Error("Client with clientId: '" + clientId + "' not found");
+        }
+        if (client.updateLastUsed) {
+          client.updateLastUsed = false;
+          this._updateLastUsed(clientId).catch(err => this.emit('error', err));
+        }
+        return client;
+      }
+    });
+    return (req) => {
+      return validator(req).then(result => {
+        if (result.status === 'auth-success') {
+          // This is only necessary if authorizedScopes or temporary credentials
+          // was used, otherwise it should already be the fixed-point.
+          // We should refactor base.API.createSignatureValidator to facilitate
+          // this... But for now this is okay...
+          result.scopes = this.resolve(result.scopes);
+        }
+        return result;
+      });
+    };
+  }
+
+  /**
+   * Remove scopes that aren't needed, e.g. if you have ["q:1", "q:*"], then
+   * the scope-set ["q:*"] is the formal-form. Basically shorter, but same
+   * level of authority.
+   */
+  static normalizeScopes(scopes) {
+    // Filter out any duplicate scopes (so we only have unique strings)
+    scopes = _.uniq(scopes);
+    // Filter out scopes that are covered by some other scope
+    return scopes.filter(scope => {
+      return !scopes.some(other => {
+        // If `scope` is `other`, then we can't filter it! It has to be
+        // strictly greater than (otherwise scopes would filter themselves)
+        if (other === other) {
+          return false;
+        }
+        // But if the other one ends with '*' and `scope` starts with its
+        // prefix then `other` is strictly greater than `scope` and we filter
+        // out `scope`.
+        return other.endsWith('*') && scope.startsWith(other.slice(0, -1));
+      });
+    });
+  }
+
+  /** Determine if scope grants a roleId, and allows owner to assume the role */
+  static grantsRole(scope, roleId) {
+    // We have 3 rules (A), (B) and (C) by which a scope may match a role.
+    // This implementation focuses on being reasonably fast by avoiding
+    // allocations whenever possible.
+
+    // Rule (A) and (B) both requires the scope to start with "assume:"
+    if (scope.startsWith('assume:')) {
+      // A) We have scope = 'assume:<roleId>', so we can assume the role
+      if (scope.length === roleId.length + 7 && scope.endsWith(roleId)) {
+        return true;
+      }
+
+      // B) guard is on the form 'assume:<prefix>*' and we have a scope on the
+      //    form 'assume:<prefix>...'. This is special rule, assigning
+      //    special meaning to '*' when used at the end of a roleId.
+      if (roleId.endsWith('*') && scope.slice(7).startsWith(roleId.slice(0, -1))) {
+        return true;
+      }
+    }
+
+    // C) We have scope as '<prefix>*' and '<prefix>' is a prefix of guard, this
+    //    is similar to rule (A) relying on the normal scope satisfiability.
+    if (scope.endsWith('*') && roleId.startsWith(scope.slice(7, -1))) {
```

Comment:
correct, I'll fix this, and implement it as two... 

We do want scope `a*` to grant all roles... you can't just break the semantics for `*` arbitrarily. It would be an absolute chaos, the end of civilization, zombies would rise and eat us all... I don't want to be blamed for that :)